util/topsql/reporter/reporter.go

```diff
@@ -0,0 +1,411 @@
+	planBinaryDecoder planBinaryDecodeFunc
+
+	// topSQLMap maps `sqlDigest-planDigest` to TopSQLDataPoints
+	topSQLMap        map[string]*topSQLDataPoints
+	MaxStatementsNum int
+
+	// normalizedSQLMap is an map, whose keys are SQL digest strings and values are normalized SQL strings
+	normalizedSQLMap map[string]string
+
+	// normalizedPlanMap is a plan version of normalizedSQLMap
+	// this should only be set from the dedicated worker
+	normalizedPlanMap map[string]string
+
+	collectCPUTimeChan chan *topSQLCPUTimeInput
+	planRegisterChan   chan *planRegisterJob
+
+	collectInterval  time.Duration
+	collectTimeout   time.Duration
+	agentGRPCAddress string
+
+	// current tidb-server instance ID
+	instanceID string
+
+	quit chan struct{}
+}
+
+// TopSQLReporterConfig is the config for TopSQLReporterImpl
+type TopSQLReporterConfig struct {
+	// PlanBinaryDecoder is used to decode the plan binary before sending to agent
+	PlanBinaryDecoder planBinaryDecodeFunc
+	// MaxStatementsNum is the capacity of the TopSQL map, above which evition should happen
+	MaxStatementsNum int
+	// CollectInterval is the interval of sending TopSQL records to the agent
+	CollectInterval time.Duration
+	// CollectTimeout is the timeout of a single agent gRPC call
+	CollectTimeout time.Duration
+	// AgentGRPCAddress is the gRPC address of the agent server
+	AgentGRPCAddress string
+	// InstanceID is the ID of this tidb server
+	InstanceID string
+}
+
+func encodeCacheKey(sqlDigest, planDigest []byte) []byte {
+	var buffer bytes.Buffer
+	buffer.Write(sqlDigest)
+	buffer.Write(planDigest)
+	return buffer.Bytes()
+}
+
+const (
+	grpcInitialWindowSize     = 1 << 30
+	grpcInitialConnWindowSize = 1 << 30
+)
+
+func newAgentClient(addr string) (*grpc.ClientConn, tipb.TopSQLAgentClient, error) {
+	dialCtx, cancel := context.WithTimeout(context.TODO(), time.Second)
+	defer cancel()
+	conn, err := grpc.DialContext(
+		dialCtx,
+		addr,
+		grpc.WithInsecure(),
+		grpc.WithInitialWindowSize(grpcInitialWindowSize),
+		grpc.WithInitialConnWindowSize(grpcInitialConnWindowSize),
+		grpc.WithDefaultCallOptions(
+			grpc.MaxCallRecvMsgSize(math.MaxInt64),
+		),
+		grpc.WithConnectParams(grpc.ConnectParams{
+			Backoff: backoff.Config{
+				BaseDelay:  100 * time.Millisecond, // Default was 1s.
+				Multiplier: 1.6,                    // Default
+				Jitter:     0.2,                    // Default
+				MaxDelay:   3 * time.Second,        // Default was 120s.
+			},
+		}),
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+	client := tipb.NewTopSQLAgentClient(conn)
+	return conn, client, nil
+}
+
+// NewTopSQLReporter creates a new TopSQL struct
+//
+// planBinaryDecoder is a decoding function which will be called asynchronously to decode the plan binary to string
+// MaxStatementsNum is the maximum SQL and plan number, which will restrict the memory usage of the internal LFU cache
+func NewTopSQLReporter(config *TopSQLReporterConfig) *TopSQLReporterImpl {
```

Comment:
Seems that this reporter is never set outside the test.

Reply:
fixed